twilio/video-quickstart-android
# Android service experiment

### Objective
* Provide a working sample of the video-quickstart-android Kotlin sample, that uses an Android Service instead of an Activity.
* Ensure users can hit "back" during a call, and the service will take over the responsibility.
* Ensure users can go back to the application and the call is restored.

### Motivation
* The idea behind this came from a client, who needs to ensure users can "minimize" the call to do other things in the app, while they are still in a call (this is a 1:1 call with another professional).
* The quicksamples are good to get started, but they fall short once you want to do more advanced stuff and the work required to separate things is quite big and prone to errors (as this PR may be!).
* It was fun to play around in 100% Kotlin ;)

### How To Test
1. Similar to the quickstart apps, fill your token in the `local.properties` file.
2. Run the app and tap "start call" 
3. Repeat 1 and 2 _with another token, on another device._
4. Once both devices are up and running, connect to the same room.
5. Once the talk is in progress, hit back in the Toolbar (or the back button in your phone). You should be prompted if you want to minimize the call. 
6. Observe: VideoActivity goes back and is destroyed, the Service (CallService) becomes Foreground (notification is shown!). The call continues. 
7. Return to the call (by tapping the notification) or the "Return to call" button.
8. Observe: Call is restored, video feed is resumed like nothing happened.
**Contributing to Twilio**

### How does this work?
 (tl;dr version)
* MainActivity is just a dummy entry point. 
* VideoActivity is the activity you'd use to display a call in progress, it binds to the Service and does all the framework stuff and views handling; it's the context in this game. It delegates to the service for any other call-related operation and passes a `CallListener` to be notified back.
* CallService is the Android Service that now hosts all the references (when possible) and takes a `CallListener` to speak with the caller. It has the ability to initiate and maintain a call, as well as to promote itself to foreground when needed. 
* CameraCapturerWrapper just wraps CameraCapturerCompat, it's not needed and should probably be simplified to an interface exposing the only needed methods. 
* I removed the code to do a network request to retrieve the token, it added noise with little gain. 

Posting this PR here in response to https://github.com/twilio/video-quickstart-android/issues/411's request ;)
> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.


### To-do(s)
Feel free to use this code as much as you want, modify, delete, etc. Keep in mind I did not write this code with the intention to open source, share, or even use this; it was more a proof of concept. 

In a production app, I'd obviously use interfaces all over the place to hide Twilio from the higher layers, since customers would likely want a less "tied" solution.
I'd delegate most of the work the service needs to do to other classes so they can be tested.
There's plenty of room for improvement elsewhere; I'd use Koin dependency injection to reduce the number of things the service needs to construct and would add plenty of Unit Test. 

There may be bugs because when something didn't really behave, I simply removed it for the sake of simplicity. 

I did talk to myself a lot during the creation of this. :)
